cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/class-sync.php

```diff
@@ -119,66 +163,654 @@ public function is_synced( $post_id ) {
 	/**
 	 * Generate a signature based on whats required for a full sync.
 	 *
-	 * @param int $post_id The post id to generate a signature for.
+	 * @param int  $attachment_id The Attachment id to generate a signature for.
+	 * @param bool $cache         Flag to specify if a cached signature is to be used or build a new one.
 	 *
 	 * @return string|bool
 	 */
-	public function generate_signature( $post_id ) {
-		$upload = $this->managers['push']->prepare_upload( $post_id );
-		// Check if has an error (ususally due to file quotas).
-		if ( is_wp_error( $upload ) ) {
-			$this->plugin->components['media']->get_post_meta( $post_id, self::META_KEYS['sync_error'], $upload->get_error_message() );
+	public function generate_signature( $attachment_id, $cache = true ) {
+		static $signatures = array(); // cache signatures.
+		if ( ! empty( $signatures[ $attachment_id ] ) && true === $cache ) {
+			$return = $signatures[ $attachment_id ];
```

Comment:
We can directly return here: `return $signatures[ $attachment_id ]`

Reply:
As I mentioned, I prefer having as few return points as possible to aid in code coverage and testing. It also helps with debugging since you know where something will return, you can track that return value to find where it changes.